### PR TITLE
Add self contained short flag

### LINKED ISSUE #19576: Enable `-s` as a convenience over `--self-contained`
This is a pretty common mode. Today, you can use `-r` to mean self-contained. We're changing that.

Proposal:

- Enable `-s` (or some other short-form like `--sc`) as a convenience over `--self-contained`.
- Use the SDK RID as an implicit RID if non is specified.

Proposed timeframe: 6.0.200.

Context: https://github.com/dotnet/designs/pull/233#discussion_r684318468

### PR DESCRIPTION
**Pull request by WeihanLi**
#19576 

Add `--sc` short for `--self-contained`

### CODE REVIEW

**Comment by richlander** [comment]
Love it. When we merge this, we should also add support an implicit RID at the same time. That will make the feature from good to ***great***.

/cc @agocke

**Comment by agocke** [comment]
I think we just need a few tests here. There are some existing tests in `GivenDotnetPublishPublishesProject.cs` that test `--self-contained` that could probably be adjusted to test this as well by adding more `InlineData` clauses

**Comment by WeihanLi** [comment]
> I think we just need a few tests here. There are some existing tests in `GivenDotnetPublishPublishesProject.cs` that test `--self-contained` that could probably be adjusted to test this as well by adding more `InlineData` clauses

Great thanks @agocke, would add some test cases

File: src/Tests/dotnet-publish.Tests/GivenDotnetPublishPublishesProjects.cs
Comment (AraHaan) [inline src/Tests/dotnet-publish.Tests/GivenDotnetPublishPublishesProjects.cs:169]: What about ``-s=true --no-self-contained``?

Also what about a shorthand for ``--no-self-contained`` as well?

File: src/Tests/dotnet-publish.Tests/GivenDotnetPublishPublishesProjects.cs
Comment (AraHaan) [inline src/Tests/dotnet-publish.Tests/GivenDotnetPublishPublishesProjects.cs:101]: What about ``-s=true``?

File: src/Tests/dotnet-publish.Tests/GivenDotnetPublishPublishesProjects.cs
Comment (WeihanLi) [reply src/Tests/dotnet-publish.Tests/GivenDotnetPublishPublishesProjects.cs:169]: I thought `-s` stands for `--self-contained true`, if `-s=false` is needed, I'd like to add more cases.

Comment (AraHaan) [reply src/Tests/dotnet-publish.Tests/GivenDotnetPublishPublishesProjects.cs:169]: It probably is, but I seen test cases where it tests ``--self-contained=true``.

File: src/Tests/dotnet-publish.Tests/GivenDotnetPublishPublishesProjects.cs
Comment (WeihanLi) [reply src/Tests/dotnet-publish.Tests/GivenDotnetPublishPublishesProjects.cs:101]: Added, thanks

File: src/Tests/dotnet-publish.Tests/GivenDotnetPublishPublishesProjects.cs
Comment (WeihanLi) [reply src/Tests/dotnet-publish.Tests/GivenDotnetPublishPublishesProjects.cs:172]: Added, thanks

Comment (richlander) [reply src/Tests/dotnet-publish.Tests/GivenDotnetPublishPublishesProjects.cs:172]: I don't think we need a short-hand for `--no-self-contained`. The plan is for that to be the default in .NET 7.

**Review by agocke — APPROVED** [review]
LGTM

**Comment by agocke** [comment]
@richlander Any more comments

**Comment by agocke** [comment]
Also @wli3 in-case I'm missing anything

**Comment by wli3** [comment]
@KathleenDollard could you double check that? We need to be careful adding new short hands, since we have only 26 letters to use

**Comment by wli3** [comment]
it is good on the technical side

**Comment by KathleenDollard** [comment]
Unfortunately, `s` is `source`. This is important on the install commands, and it also works on `restore`. 

We avoid having the same abbreviation mean different things, especially for closely related commands. And we work hard to have consistent abbreviations where a particular option is available. `build` and `restore` are obviously closely linked commands and it would seem likely consider them sharing more options in the future. 

Since `build` does a `restore` and supports `s`, taking this would mean that we could not add `--source` to pass through to restore. It would also that we could not specify `--self-contained` for `restore` (to get the right packages). Unfortunately, `c` doesn't work and `-sc` would be confusing for Posix users. We generally avoid two letter abbreviations because of that, but as Will points out, there are only 26 letters and there is value to an abbreviation here.

**Comment by AraHaan** [comment]
I think there is no choice but to abbreviate this as ``-sc``.

**Comment by agocke** [comment]
We could do `--sc`, which is still short but also fits posix convention

**Comment by richlander** [comment]
`--sc` sounds great to me.

I still want to hold on this change until we add an implicit RID by default.

**Comment by AraHaan** [comment]
> `--sc` sounds great to me.
> 
> I still want to hold on this change until we add an implicit RID by default.

Is there a way to do that based on what it detects on the os and what cpu bitness it detects as well?

**Comment by WeihanLi** [comment]
Thanks, will update to `--sc` later. 

@richlander as for the default implicit RID, I'm going to implement it in another PR <https://github.com/dotnet/sdk/pull/19725>, I'm not sure about the changes, maybe @agocke could help?

**Comment by wli3** [comment]
FYI @KathleenDollard is not available today.

**Comment by marcpopMSFT** [comment]
Got @KathleenDollard's signoff offline.  @wli3 , anything else before we approve and merge?

**Comment by AraHaan** [comment]
Is this ready to merge yet?

**Review by wli3 — APPROVED** [review]

**Pull request merged**
